Fix loading collapsed tabs and legacy session support

toofar opened the pull request:
Relating to https://github.com/qutebrowser/qutebrowser/issues/8076 and following on from https://github.com/qutebrowser/qutebrowser/commit/a2f98aa78b529de9a730beaafe522b38a8a977c3 this PR proposes a fix and an improvement:

**fix loading of tab in collapsed tab groups**

Tabs in collapsed tab groups were being dropped because the code was iterating over the tabs loaded into the tab widget instead of all those in the tree. Easy spot, easy fix.

**support loading legacy tree tab session formats**

There are some long time community members who have been using this PR for a while and may have built up complicated session files which they are dreading having to rebuild. Currently if they pull the latest on the `tree-tab-integration` branch the browser will just crash on startup.

I had a look at how hard it would be to continue supporting the previous session format and it turned out to be pretty easy by copying some old code and adding a couple of conditionals. A few two many conditionals for my tastes but at least they are tiny.

**bonus: remove unused attribute**

Tree tabs were ending up in session files with their uid in both the `node` and `uid` attribute. It looks like only the `uid` one is being used.

**PS**: I haven't reviewed a2f98aa78b529de9a yet, but I've tested it! Everything seems to be working as it was. And from a brief look the diff sure looks smaller. I think `_save_all()` looks much cleaner iterating on tabs like normal and then taking a quick diversion to do add tab data
**PPS**: I've seen another issue where if you load a non-tree session while sessions are enabled every tab will be opened as "related" so instead of a flat set of tabs loaded you get a staircase. This might depend on what your `new_position` settings are set to. This could be fixed by setting `related=True` in the call to `tabopen()` but I think this is exposing a problem you can run into with the existing session loading too. Eg tabopen has `related=True` false and that isn't overriden by the session code and the session code passes `background=False`. Both of which seem incorrect, so it seems the session code is using knowledge of the positioning code in tabbedbrowser instead of overriding it completely (eg `background=False, related=False, idx=i`). So since I think it's a preexisting thing I would prefer to wait until we fix up the tests around sessions (lots of them are disabled on webengine) before fixing it "properly". But if someone complains I suppooose fixing it in an isolated manner is okay. ~TODO: raise a ticket to track this~ resolved in this PR
Edit: this also causes issues with:
```
- about:blank?one
  - about:blank?two
    - about:blank?three (collapsed) (active)
      - about:blank?four
- about:blank?five
```
If you load the session tab five is also collapsed. If you toggle collapse it gets restored to the correct visibility. This seems to be happening in `_position_tab()` because `related` is set to True. Not sure why it's only happening when there is a collapsed tab, or why the `child.parent = root_node` at the end of `_load_tree()` isn't fixing it. But it gets fixed with related=False anyhow...
Hmm, had a look at the e2e tests, which do work on webengine, they don't reproduce this issue because they test against a saved session and the session saves fine, it's the tabs in the browser that are borked (how!). Well, session-load does have some coverage so might just have to be a "it doesn't make things worse" change. Either that or we add a `:debug-dump-tree $winid` command.
Anyway, you can reproduce manually with `python3 -m qutebrowser -T -s tabs.tree_tabs true -s tabs.position left ":open about:blank?one" ":open -tr about:blank?two" ":open -tr about:blank?three" ":open -t about:blank?four" ":tab-focus 2" ":tree-tab-toggle-hide" ":cmd-later 300 session-save asdf" ":cmd-later 350 session-load asdf"` then look how one window looks like it has less tabs than the other one!

pinusc reviewed (commented):
Tests always sound like a good idea.

I don't like the idea of `.widgets()` returning things that aren't visible. Maybe adding a method `.tabs()` to `TabbedBrowser` that is just an alias of `widgets`, overridden in `TreeTabbedBrowser` to return all tabs, including collapsed. I don't think there's any places in treetabs where `.widgets()` explicitly expects only visible tabs, but this half-rename would also ensure compatibility. Maybe that's overcomplicating things.

pinusc reviewed (commented):
I did think people might be unhappy with the breaking change :worried: 

I didn't think legacy code would be great to lug around, but I'm not the maintainer here... that being said, we could keep this legacy code for now, adding a deprecation prompt, and then removing it altogether before merging into main. Loading the session with this code and then immediately saving it should "update" the old session file. This gives users a chance to update their sessions, essentially serving as a conversion script of sorts; but then we don't have to keep legacy code in.

toofar commented:
> I don't like the idea of .widgets() returning things that aren't visible. Maybe adding a method .tabs() to TabbedBrowser that is just an alias of widgets, overridden in TreeTabbedBrowser to return all tabs, including collapsed. 

Yeah, good point. And thanks for engagement with a comment buried in a commit message! 

I wanted to have a look at what existing places in the code would do if `TabbedBrowser.widgets()` started returning hidden tabs. I used semgrep to search for `$OBJECT.widgets()` and `$OBJECT.widget.count()`. I came up with the following classification, as in "could the calling code handle invisible tabs":

    shouldn't (it would break)
    11 instances
    could- (with changes, but no change in behaviour)
    5 instances
    could+ (with changes, could be an enhancement)
    1 instances
    should (maybe we should fix this, I added it to https://github.com/orgs/qutebrowser/projects/3/views/2?pane=issue&itemId=52187228)
    1 instances
    can (eg if widgets() started returning them now)
    4 instances

So the majority of places where tabs are iterated over or counted at the moment are specifically to do with moving or focusing tabs and wouldn't deal well with tabs that weren't in the tab widget as things are.

<details>
<summary>
Full notes from those greps:
</summary>

```
qutebrowser/browser/commands.py
      899┆ for i, tab in enumerate(self._tabbed_browser.widgets()):

      Iterates through all tabs to close them. Works fine with tree tabs
      because the top tab of a collapsed group will always be in the widget.
      Does have different behaviour if you have a collapsed tab vs not a
      collapsed tab though. If you have selected an uncollapsed tab with
      children and run :tab-only you are left with only the selected tab left.
      If you have selected a collapsed tab with children, you are left with
      the tab and it's children left. I think there is an argument made that
      that should be consistent in either directing but I don't think that's a
      blocker (it's not clear to me what the correct thing to do is anywhow).

      Can handle hidden tabs: not as it is currently, `TabbedBrowser._remove_tab()`
      calls `self.widget.indexOf()` and complains if it doesn't find it.
      Classification: shouldn't

qutebrowser/browser/qutescheme.py
      190┆ for tab in tabbed_browser.widgets():
      
      Intent: for qute://tabs
      Can handle hidden tabs: yes
      Classification: can

qutebrowser/browser/webengine/notification.py
      361┆ for idx, tab in enumerate(tabbedbrowser.widgets()):

      Intent: when someone clicks on a notification we try to match the url on
        the notifications with a tab to focus.
      Can handle hidden tabs: not as it is currently, would have to learn to
        un-collapse a tab group, and probably make it obvious that that
        happened. Would be an interesting to figure out what affordance we
        could give the user that "a tab in this collapsed group wants focus",
        but probably something for another day.
      Classification: could+

qutebrowser/completion/models/miscmodels.py
      202┆ tab_titles = (tab.title() for tab in tabbed_browser.widgets())

      Intent: lists tab titles in the :tab-give completion so you get a better
        idea of the window you are going to be sending a tab too. 
      Can handle hidden tabs: yes, but probably don't need hidden tabs showing
        here. Although I don't undestand why this is shoving all of the tab
        titles on to the end of the category name anyway? Seems really hard to
        read, maybe a hack to allow completing on tab titles? Hmmm
      Classification: can

qutebrowser/mainwindow/tabbedbrowser.py
      425┆ for idx, tab in enumerate(reversed(self.widgets())):
      
      Intent: close all the tabs
      Can handle hidden tabs: same as tab-only, not currently
      Classification: shouldn't

      738┆ for tab in self.widgets():

      Intent: update favicons for all loaded tabs
      Can handle hidden tabs: nope
      Classification: shouldn't

qutebrowser/misc/crashsignal.py
      120┆ for tab in tabbed_browser.widgets():

      Intent: save open URLs on crash for the purpose of restoring them
      Can handle hidden tabs: yes, and probably should!
      Classification: should

qutebrowser/app.py
      310┆ if cur_window.tabbed_browser.widget.count() == 0:

      Intent: see if there are any tabs open
      Can handle hidden tabs: yes
      Classification: can

qutebrowser/browser/commands.py
       63┆ return self._tabbed_browser.widget.count()

       Aliased to self._count(), used in:

       def tab_give:
       Intent: see if we are going to remove the last tab
       Can handle hidden tabs: yes, but doesn't need to as the last tab will
         aways be visible. As a note this check is broken with --recursive
       Classification: could-
       
       def tab_prev and tab_next:
       Intent: check if we have any tabs open
       Can handle hidden tabs: no (can't focus them), and shouldn't need to
       Classification: shouldn't

       def tab_focus:
       Intent: to validate the index
       Can handle hidden tabs: no (can't focus them), and shouldn't need to
       Classification: shouldn't

       def tab_move:
       Intent: to validate and wrap the index
       Can handle hidden tabs: no, the current logic ignores them, which is
         probably fine
       Classification: shouldn't

       def _cntwidget()
       Intend: return a widget by index
       Can handle hidden tabs: no
       Classification: shouldn't

      1068┆ if not 0 < idx <= tabbed_browser.widget.count():

      Intent: validate tab index for tab_select and tab_focus
      Can handle hidden tabs: no
      Classification: shouldn't

qutebrowser/commands/command.py
      377┆ elif 1 <= self._count <= tabbed_browser.widget.count():

      Intent: get tab with count for commands
      Can handle hidden tabs: no
      Classification: shouldn't

qutebrowser/completion/models/miscmodels.py
      127┆ for idx in range(tabbed_browser.widget.count()):

      Intent: get tabs for completions
      Can handle hidden tabs: no, the commands being completed likely wouldn't
        handle them
      Classification: shouldn't

qutebrowser/mainwindow/mainwindow.py
      647┆ tab_count = self.tabbed_browser.widget.count()

      Intent: see if there are any tabs open before shutting down
      Can handle hidden tabs: yeah, but doesn't need to
      Classification: could-

qutebrowser/mainwindow/tabbedbrowser.py
      254┆ return utils.get_repr(self, count=self.widget.count())

      Intent: add count of open tabs to repr()
      Can handle hidden tabs: yeah
      Classification: can

      291┆ for i in range(self.widget.count()):

      Intent: return open tabs
      Can handle hidden tabs: recursion error
      Classification: shouldn't

      459┆ count = self.widget.count()

      Intent: see if there are any tabs left open after closing
      Can handle hidden tabs: yeah, but doesn't matter, only cares about top
        level tabs which can't currently be hidden
      Classification: could-

      537┆ only_one_tab_open = self.widget.count() == 1

      Intent: see if we only have one tab open
      Can handle hidden tabs: yeah, but doesn't matter, only cares about top
        level tabs which can't currently be hidden
      Classification: could-

      649┆ if config.val.tabs.tabs_are_windows and self.widget.count() > 0:

      Intent: check to see if the window is empty or not
      Can handle hidden tabs: yeah, but doesn't matter, only cares about top
        level tabs which can't currently be hidden
      Classification: could-

      677┆ self.widget.count())

      Intent: emit tab_index_changed on openeing a background tab
      Can handle hidden tabs: not with current logic, I think it's a pretty
        tab widget specific signal
      Classification: shouldn't

```
</details>

So if there are valid use cases for iterating over all tabs (including hidden ones) and for iterating only over visible (or focusable?) tabs, what would APIs look like that make it clear to callers clear how to do each without making them deal with concepts they don't need to? As a thought experiment: when we put tabs and windows into the extension API what should that look like?

If we add a `.tabs()` method that returns all tabs, even hidden ones, what are the chances of callers who are doing tab widget related stuff calling that by mistake? Knowing when to call `.tabs()` and when to call `.widgets()` might not be obvious.

What about something like `.tabs(include_hidden=True)`? That way there's only one method to call and it's hopefully obvious how to make it do what you want. And then default `include_hidden` to `False` because that the most common use case currently.

A related problem is that once we have an API that can return a mix of tabs which are in the tab widget and hidden tabs, how can the callers know which of those tabs was hidden? There are isHidden() and isVisible() QWidget methods but those are set the same for hidden and background tabs. I kinda want to add a tab.idx() which will return the index within the window, or None. That would also be useful for all the places that call browser.widget.indexOf(tab) to not have to get a reference to the browser. Anyway, that's probably a problem for another day.

> we could keep this legacy code for now

I'm thinking we just remove it before merging to main and provide a userscript that'll update any sessions people have that they haven't bothered to upgrade yet. Maybe even point to the userscript in an error message if we can detect those session easy enough.

Or since no-one has complained all week maybe no-one is using it :shrug:

toofar commented:
Alright, I made some more changes:

* rebased so that tests and lint can pass
* spotted another issue where we forgot to focus the relevant tab after loading a tree tab window
* added a new `TreeTabbedBrowser.tabs()` method, as discussed. See the commit message for a discussion on editing the parent class for stuff it doesn't implement vs requiring callers to know about the child class
* fixed the issue I was rambling about at the bottom of the PR description where a tab could be hidden when it shouldn't be. Still not 100% sure what chain of events leads to it, it might be some kind of timing issue as nodes' parents get updated a lot when a session is being loaded. But it's fixed by setting `related=False` anyway, and I managed to find a reproducer for a related issue with "flat tabs" too to make a convincing test case. I suspect we should be skipping `TreeTabbedBrowser._position_tab()` when an index is passed to `tabopen()` but I think I have a suggestion for refactoring that on another issue anyway, hopefully there will be more test scaffolding in place by then. (ref: https://github.com/qutebrowser/qutebrowser/pull/8083)

Anyway, I think this PR has grown in scope enough. The main thing was fixes around session loading and I've managed to cover three of them. I've squeezed in legacy session support and a new API method in here too, the latter was particularly interesting.

Lemme know if you have time to review in the next couple of days. Otherwise I'll go ahead and merge into the integration branch, you are of course free to comment on the PR after I've merged it to raise things that could be better!

toofar commented:
Almighty, merging now. To close this out I had another look at the previous changes to session loading and have a couple more notes on that, but nothing here has any priority over the other stuff we already have to look at I think. And my overall opinion from earlier that the new session loading stuff is cleanly done still stands, it re-uses all of the existing session loading and just adds a couple of diversions to load or save tree specific data.

```patch
diff --git a/qutebrowser/misc/sessions.py b/qutebrowser/misc/sessions.py
index dd63904cdc4..fe8e7ba24d8 100644
--- a/qutebrowser/misc/sessions.py
+++ b/qutebrowser/misc/sessions.py
@@ -281,13 +317,32 @@ def _save_all(self, *, only_window=None, with_private=False, with_history=True):
             if getattr(active_window, 'win_id', None) == win_id:
                 win_data['active'] = True
             win_data['geometry'] = bytes(main_window.saveGeometry())
-            win_data['tabs'] = []
             if tabbed_browser.is_private:
                 win_data['private'] = True
+
+            win_data['tabs'] = []
???          ^ not clear why this was moved
             for i, tab in enumerate(tabbed_browser.widgets()):
                 active = i == tabbed_browser.widget.currentIndex()
-                win_data['tabs'].append(self._save_tab(tab, active,
-                                                       with_history=with_history))
+                tab_data = self._save_tab(tab,
+                                          active,
+                                          with_history=with_history)
???            I wonder if we could have _save_tab() fill in the tree data if tab.node was set?
???            For the distant future when refactoring the sessions code (there should be an issue for it somewhere) it might be nice if we could make AbstractTab and MainWindow handle their own serialization
+                if tabbed_browser.is_treetabbedbrowser:
+                    node = tab.node
+                    node_data = {
+                        'node': node.uid,
+                        'parent': node.parent.uid,
+                        'children': [c.uid for c in node.children],
+                        'collapsed': node.collapsed,
+                        'uid': node.uid
+                    }
+                    tab_data['treetab_node_data'] = node_data
+                win_data['tabs'].append(tab_data)
+            if tabbed_browser.is_treetabbedbrowser:
+                root = tabbed_browser.widget.tree_root
+                win_data['treetab_root'] = {
+                    'children': [c.uid for c in root.children],
+                    'uid': root.uid
+                }
             data['windows'].append(win_data)
         return data
 
@@ -455,6 +510,45 @@ def _load_tab(self, new_tab, data):  # noqa: C901
         except ValueError as e:
             raise SessionError(e)
 
+    def _load_tree(self, tabbed_browser, tree_data):
+        tree_keys = list(tree_data.keys())
+        if not tree_keys:
+            return None
???      ^ in what situation could this happen?
+
+        root_data = tree_data.get(tree_keys[0])
+        if root_data is None:
+            return None
???      ^ same here, shouldn't this be an error? If this is something we don't expect to happen I would rather complain about it loudly if it doesn't instead of silently ignoring it
+
+        root_node = tabbed_browser.widget.tree_root
+        tab_to_focus = None
+        index = -1
+
+        def recursive_load_node(uid):
+            nonlocal tab_to_focus
+            nonlocal index
+            index += 1
+            tab_data = tree_data[uid]
+            node_data = tab_data['treetab_node_data']
+            children_uids = node_data['children']
+
+            if tab_data.get('active'):
+                tab_to_focus = index
+
+            new_tab = tabbed_browser.tabopen(background=False)
+            self._load_tab(new_tab, tab_data)
+
+            new_tab.node.parent = root_node
???         ^ I assume this is being done to avoid an error in the tree walking code further down somewhere. Probably could pass the parent as an arg into recursive_load_node() and then set the real parent here rather than setting a placeholder one and then setting the real one after the function returns
+            children = [recursive_load_node(uid) for uid in children_uids]
+            new_tab.node.children = children
+            new_tab.node.collapsed = node_data['collapsed']
+            return new_tab.node
+
+        for child_uid in root_data['children']:
+            child = recursive_load_node(child_uid)
+            child.parent = root_node
+
+        return tab_to_focus
+
     def _load_window(self, win):
         """Turn yaml data into windows."""
         window = mainwindow.MainWindow(geometry=win['geometry'],
```
